programs/cleos/main.cpp

```diff
@@ -1725,6 +1752,86 @@ int main( int argc, char** argv ) {
 
+   // convert subcommand
+   auto convert = app.add_subcommand("convert", localized("Pack and unpack transactions"), false); // TODO also add converting action args based on abi from here ?
+   convert->require_subcommand();
+
+   // pack transaction
+   string plain_signed_transaction_json;
+   auto pack_transaction = convert->add_subcommand("pack_transaction", localized("From plain signed json to packed form"));
+   pack_transaction->add_option("transaction", plain_signed_transaction_json, localized("The plain signed json (string)"))->required();
+   pack_transaction->set_callback([&] {
+      fc::variant trx_var;
+      try {
+         trx_var = json_from_file_or_string(plain_signed_transaction_json);
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Fail to parse plain transaction JSON '${data}'", ("data", plain_signed_transaction_json))
+      signed_transaction trx = trx_var.as<signed_transaction>();
+      std::cout << fc::json::to_pretty_string(fc::variant(packed_transaction(trx, packed_transaction::none))) << std::endl;
+   });
+
+   // unpack transaction
+   string packed_transaction_json;
+   bool unpack_action_data_flag = false;
+   auto unpack_transaction = convert->add_subcommand("unpack_transaction", localized("From packed to plain signed json form"));
+   unpack_transaction->add_option("transaction", packed_transaction_json, localized("The packed transaction json (string containing packed_trx and optionally compression fields)"))->required();
+   unpack_transaction->add_flag("--unpack-action-data", unpack_action_data_flag, localized("Upack all action datas within transaction and only return those, needs interaction with nodeos"));
+   unpack_transaction->set_callback([&] {
+      fc::variant trx_var;
+      try {
+         trx_var = json_from_file_or_string(packed_transaction_json);
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Fail to parse packed transaction JSON '${data}'", ("data", packed_transaction_json))
+      string trx_hex;
+      try {
+         trx_hex = trx_var["packed_trx"].as_string();
+      } EOS_RETHROW_EXCEPTIONS(transaction_type_exception, "Missing packed_trx field in provided JSON '${data}'", ("data", packed_transaction_json))
+
+      std::vector<char> trx_blob(trx_hex.size()/2);
+      fc::from_hex(trx_hex, trx_blob.data(), trx_blob.size());
+      transaction unpacked_trx = fc::raw::unpack<transaction>(trx_blob);
+      // TODO would it be nice to put the unpacked_trx actions into the trx actions in a separate unpacked_data field  or somesuch ?
+      if (unpack_action_data_flag) {
+         for ( const auto& a : unpacked_trx.actions ) {
+            fc::variant unpacked_action_data = bin_to_variant(a.account, a.name, a.data);
+            std::cout << fc::json::to_pretty_string(unpacked_action_data) << std::endl;
+         }
+      } else {
+	 // TODO should the "sigantures" and "context_free_data" fields be copied from trx_var into the action field(s) ?
```

Comment:
yes, signatures and context_free_data should be added to the signed transaction if available.